teachertool/src/types/criteriaParameters.ts

```diff
@@ -0,0 +1,139 @@
+import { Strings } from "../constants";
+import { logError } from "../services/loggingService";
+import { ErrorCode } from "./errorCode";
+
+export type CriteriaParameterType = "string" | "longString" | "number" | "block" | "system";
+
+export interface CriteriaParameterValidationResult {
+    valid: boolean;
+    message?: string;
+}
+
+// Represents a parameter definition in a catalog criteria.
+export class CriteriaParameter {
+    name: string;
+    type: CriteriaParameterType;
+    default: string | undefined;
+    paths: string[]; // The json path(s) to update with the parameter value in the catalog criteria.
+
+    constructor(name: string, type: CriteriaParameterType, paths: string[], defaultValue?: string) {
+        this.name = name;
+        this.type = type;
+        this.default = defaultValue;
+        this.paths = paths;
+    }
+
+    validate(value: string): CriteriaParameterValidationResult {
+        return { valid: true };
+    }
+}
+
+export class BaseStringParameter extends CriteriaParameter {
+    maxCharacters: number | undefined;
+
+    constructor(
+        name: string,
+        type: CriteriaParameterType,
+        paths: string[],
+        defaultValue?: string,
+        maxCharacters?: number
+    ) {
+        super(name, type, paths, defaultValue);
+        this.maxCharacters = maxCharacters;
+    }
+
+    override validate(value: string): CriteriaParameterValidationResult {
+        if (!value) return { valid: true }; // Unset is okay for initial value
+
+        if (this.maxCharacters && value.length > this.maxCharacters) {
+            return { valid: false, message: Strings.ExceedsMaxLength };
+        }
+        return { valid: true };
+    }
+}
+
+export class StringParameter extends BaseStringParameter {
+    constructor(template: CriteriaParameter, maxCharacters?: number) {
+        super(template.name, "string", template.paths, template.default, maxCharacters);
+    }
+}
+
+export class LongStringParameter extends BaseStringParameter {
+    constructor(template: CriteriaParameter, maxCharacters?: number) {
+        super(template.name, "longString", template.paths, template.default, maxCharacters);
+    }
+}
+
+export class NumberParameter extends CriteriaParameter {
+    min: number | undefined;
+    max: number | undefined;
+
+    constructor(template: CriteriaParameter, min?: number, max?: number) {
+        super(template.name, "number", template.paths, template.default);
+        this.min = min;
+        this.max = max;
+    }
+
+    override validate(value: any): CriteriaParameterValidationResult {
+        // Ensure the value is numeric and within the specified range.
+        const num = Number(value);
+        if (isNaN(num)) {
+            return {
+                valid: false,
+                message: Strings.MustBeANumber,
+            };
+        }
+        if (this.min && num < this.min) {
```

Comment:
Did you want to support a min value of zero here?

Reply:
Ah good catch